### ISSUE TITLE
Batch Normalization for RNN

### ISSUE THREAD

**Issue by trungnt13** [Lasagne/Lasagne#577]
The paper: [Deep Speech 2](http://arxiv.org/abs/1512.02595) used _sequence-wise_ normalization for recurrent computation which was proved to substantially improves final generalization error while greatly accelerating training, very deep networks of RNNs on large datasets.
![BatchNorm](http://www.sciweavers.org/upload/Tex2Img_1452875406/render.png)

My understand of this idea is applying batch normalization in _precompute_input_ code:
```
        if self.precompute_input:
            input = _batch_norm(T.dot(input, W_in_stacked))
```
However, if  we set self.precompute_input=False, it would be very inefficient to apply batch_norm every time-step.
Can you suggest better approach for this batch normalization?

Discussion about batch norm for GRU and LSTM on reddit: https://www.reddit.com/r/MachineLearning/comments/4082yn/rnn_batch_normalization_for_grulstm/

**Comment by f0k** [comment]
> My understand of this idea is applying batch normalization in precompute_input code:
> ```python
> if self.precompute_input:
>     input = _batch_norm(T.dot(input, W_in_stacked))
> ```

Correct. To perform sequence-wise normalization, you'll need to pass `axes=(0, 1)`, so it normalizes over both the mini batch items and time steps.

> However, if we set `self.precompute_input=False`, it would be very inefficient to apply batch_norm every time-step.

Without `precompute_input`, it's not possible to apply sequence-wise batch normalization. You need to have the results of the input-to-hidden transformation for the full batch and over the full sequence. If you compute it time step by time step, you don't have the necessary statistics available (it's possible to normalize using the statistics from what you've seen so far in a sequence, but this doesn't work as well, I think they tried in http://arxiv.org/abs/1510.01378).

----

To make it easy to apply sequence-wise batch normalization in a Lasagne RNN, we'll need to implement #425. Right now you can only hack it into the RNN layers.

**Comment by MarcoCiccone** [comment]
I'm trying to modify the `GRULayer` as you stated above, but `batch_norm()` requires a Layer object in input. Am I right?  
How can I use `batch_norm` function inside the GRULayer?
Maybe I'm missing something..

**Comment by f0k** [comment]
> How can I use `batch_norm` function inside the GRULayer?

You're right, it's implemented as a layer. If you just need the Theano expression, you can use the following pattern:
```python
def __init__(self, incoming, ..., **kwargs):
    ...
    self.bn = BatchNormLayer(the_shape)  # create BN layer for correct input shape
    self.params.update(self.bn.params)  # make BN params your params
def get_output_for(self, input, **kwargs):
    data = ...  # your Theano expression
    data = self.bn.get_output_for(data, **kwargs)
    ...
```
That is, you create a BatchNormLayer in your constructor, passing it the input shape of the data you'll want to normalize later. In your custom layer's `get_output_for()`, you call the BN layer's `get_output_for()` to normalize the data.

**Comment by MarcoCiccone** [comment]
Thank you for your help! It was exactly what I was looking for

**Comment by avostryakov** [comment]
Thank you, f0k for the complete description! By the way, when I implemented LSTM class with batch normalization inside I noticed that comments near precompute_input block have errors. I created a pull request: https://github.com/Lasagne/Lasagne/pull/634

**Comment by pedromnasc** [comment]
Here are the new ideas about batch normalization for RNNs: http://arxiv.org/pdf/1603.09025v1.pdf

**Comment by MarcoCiccone** [comment]
very interesting! but how this could be implemented in Lasagne? it seems that each gate needs it's BN for Wx and Wh separately but what does it mean: "Consequently, we recommend using separate statistics for each timestep to preserve information of the initial transient phase in the activations." ? 

Which is the axis that have to be batch-normalized?

**Comment by f0k** [comment]
> but how this could be implemented in Lasagne?

Currently, it requires writing a custom `LSTMLayer` class.

> Which are the axis that have to be batch-normalized?

To have separate statistics for each timestep and feature, you need to batch-normalize with `axes=(0,)`.

**Comment by avostryakov** [comment]
I implemented Recurrent Batch Normalization layer for LSTM from the article above. This is the code: https://gist.github.com/avostryakov/2e6d2b0055bc0d6fdbbf6d8d5b4ec303

I used usual BatchNormLayer to normalize input parameters but normalization of hidden parameters is a little tricky. Implementation of this is a little ugly because I didn't find a better way to update statistics (running_mean and running_inv_std inside scan step function). So, I created python list of shared variables with length equals a count of time steps (lines 45-49). It compiled and works some way.

@f0k First of all, can you look at the implementation. Maybe there is more efficient way to implement the article. 

At second, I have a question: why in BatchNormLayer it is used cloning of statistics variables:
running_mean = theano.clone(self.mean, share_inputs=False)
running_inv_std = theano.clone(self.inv_std, share_inputs=False)

In my implementation, it works without cloning. I didn't find in Theano doc why we should use clone function.

**Comment by f0k** [comment]
> At second, I have a question: why in BatchNormLayer it is used cloning of statistics variables:
running_mean = theano.clone(self.mean, share_inputs=False)
running_inv_std = theano.clone(self.inv_std, share_inputs=False)

Because if we attach the default_updates directly to self.mean and self.inv_std, they will stick and also be applied during testing.

> In my implementation, it works without cloning.

I can't check your implementation right now, but if you're sure the statistics are not updated during testing now, it may have to do with `theano.scan()`. I wouldn't be sure that default_updates inside the step function are included.

**Comment by avostryakov** [comment]
It works in unroll_scan=True mode. This code will be not work because I can't pass python lists in theano.scan in seqences attribute. And I'm sure that statistics will ne updated during testing because deterministic=True in this case and another piece of of code will be executed.

**Comment by avostryakov** [comment]
But what is interesting: the article was implementated on theano too. And I don't know how they updated statistics in RNN loop for a long sequencies. For sequenced MNIST there are 768 time steps. Maybe they didn't use theano.scan too.

**Comment by f0k** [comment]
> And I'm sure that statistics will not be updated during testing because deterministic=True in this case and another piece of code will be executed.

If you call it with deterministic=False once, then the default_updates attribute would be set on the shared variables *permanently*, so calling it with deterministic=True afterwards (within the same script) would still include the updates. That's why I had to clone the shared variables instead of using the original ones.

**Comment by avostryakov** [comment]
Aha! I see now!

**Comment by avostryakov** [comment]
@f0k By the way, we can set the parameter no_default_updates=False for a test function. In this case, shared_variable.defaul_update will not be applied. So, It looks like we don't need to clone shared variables. What do you think?

**Comment by f0k** [comment]
> What do you think?

This would require users to think of another flag to set, and it would prevent users from using `default_updates` for anything else (it's useful for random variables in RNNs as well, for example). So no, cloning the shared variables is the best solution unless we implement #429. It doesn't cost anything either, so why are you looking for an alternative? :)

**Comment by avostryakov** [comment]
I'm just trying to implement the article above (Recurrent Batch Normalization) as simple and efficient way as possible. I have spoken with Noize here: https://github.com/Theano/Theano/issues/4386 and I'm not sure that is possible to add cloning in all of this implementation. I will try tomorrow.

**Comment by avostryakov** [comment]
@f0k Thank you very much for your help with Recurrent Batch Normalization in theano issue above! it works! I'll try to convert it in default_update to encapsulate it in BatchNormalizedLSTM Layer.

**Comment by avostryakov** [comment]
@f0k My final implementation of Recurrent Batch Normalization with cloning statistics and using default_update attribute: https://gist.github.com/avostryakov/2e6d2b0055bc0d6fdbbf6d8d5b4ec303

From my point of view it works. Can you check it when you will have time?

**Comment by f0k** [comment]
> From my point of view it works. Can you check it when you will have time?

I don't have a lot of time these days, unfortunately... can you reproduce the original paper's results for some task? If you, we'd welcome you submitting it to [Lasagne/Recipes](/lasagne/recipes)!

**Comment by avostryakov** [comment]
I'll try to reproduce results from the paper.

**Comment by joelmoniz** [comment]
About Recurrent Batch Normalization, this is the repository with an implementation in Theano by the authors of the paper: https://github.com/cooijmanstim/recurrent-batch-normalization

**Comment by TigranGalstyan** [comment]
@avostryakov Any news?

**Comment by avostryakov** [comment]
@TigranGalstyan I haven't done it. It looks like residual connections solve the same task as Recurrent Batch Normalization but easier to implement and train.

**Comment by webeng** [comment]
@avostryakov Are you aware of any implementation of residual connections?

**Comment by avostryakov** [comment]
@webeng residual connections are just a summation of two tensors. The first tensor is input to the N-th layer, the second tensor is input to the N+1 layer. Nothing to implement.

**Comment by f0k** [comment]
> From my point of view it works. Can you check it when you will have time?

I don't really have time, but a colleague (@neosatrapahereje) also asked me about this on Friday and it just came up on the mailing list: https://groups.google.com/forum/#!topic/lasagne-users/EczUQckJggU
So I finally took a closer look at your implementation and found the following:

* https://gist.github.com/avostryakov/2e6d2b0055bc0d6fdbbf6d8d5b4ec303#file-batch_normalized_lstm_layer-py-L44-L45 applies batch normalization over axis 0, but inside the recurrent layer that's the sequence dimension, not the batch dimension. You would need to translate the `batch_axes` argument to swap 0 and 1. And IIRC, in the paper they normalize the input-to-hidden activations across both the sequence and the batch.
* https://gist.github.com/avostryakov/2e6d2b0055bc0d6fdbbf6d8d5b4ec303#file-batch_normalized_lstm_layer-py-L101 misses to pass on `deterministic` and `**kwargs` to `get_output_for()`.
* The way the running averages are computed in https://gist.github.com/avostryakov/2e6d2b0055bc0d6fdbbf6d8d5b4ec303#file-batch_normalized_lstm_layer-py-L121-L128 seems to be inefficient. You could have `scan` return the unnormalized hidden states as an extra output and then compute the running average updates outside scan in a single operation, instead of stacking all those set_subtensor Ops.
* The code misses the extension to test sequences longer than the training sequences (but of course that's not an error, just a bonus feature).

**Comment by vincent341** [comment]
Thanks very much for your suggestions. For the second suggestion , do you mean I should modify [line 101](https://gist.github.com/avostryakov/2e6d2b0055bc0d6fdbbf6d8d5b4ec303#file-batch_normalized_lstm_layer-py-L101) like this?

```python
input = self.bn_input.get_output_for(T.dot(input, W_in_stacked),deterministic=True) + b_stacked
```
Is that right?
